src/System.Private.CoreLib/shared/System/Math.cs

```diff
@@ -165,22 +165,40 @@ public static double BitIncrement(double x)
 
         public static unsafe double CopySign(double x, double y)
         {
-            // This method is required to work for all inputs,
-            // including NaN, so we operate on the raw bits.
+            const long signMask 
+                = unchecked((long)0b_1000_0000__0000_0000__0000_0000__0000_0000__0000_0000__0000_0000__0000_0000__0000_0000);
 
-            long xbits = BitConverter.DoubleToInt64Bits(x);
-            long ybits = BitConverter.DoubleToInt64Bits(y);
+            if (Sse.IsSupported)
+            {
+                // Create vectors of the elements, and make them float to allow using SSE rather than SSE2 instructions
+                var xvec = Vector128.CreateScalarUnsafe(x).AsSingle();
+                var yvec = Vector128.CreateScalarUnsafe(y).AsSingle();
 
-            // If the sign bits of x and y are not the same,
-            // flip the sign bit of x and return the new value;
-            // otherwise, just return x
+                // Remove the sign from x, and remove everything but the sign from y
+                // Creating from a 'const long' is better than from the correct 'const float/double'
```

Comment:
What was the performance difference between using `Vector128.CreateScalarUnsafe(signMask).AsSingle()`, which generates:
```
mov      rax, 0x8000000000000000
vmovd    xmm2, rax
```

and using `Vector128.CreateScalarUnsafe(-0.0);`, which generates:
```
vmovsd xmm0, dword [rip+0x15]
```

The latter has the load from memory, but smaller codegen, which I believe could be impactful to tight loops.
MSVC, GCC, and Clang all look to prefer the latter scenario...

Reply:
I tried both and in the naive-ish simple creation benchmark, the first appeared to be marginally faster so I chose it

Reply:
Also, now `ANDNPS` is used, there is only 1 load rather than 2

Reply:
> the JIT doesn't seem to inline this method in its current state

This is probably because the IL for all code paths is sufficiently big to skip the initial heuristics.

What we've done for some of the other code is mark the method as `AggressiveInlining` and keep any code-paths we know are small enough inline (i.e. generally the intrinsic paths). The software fallback (or other paths which might not be "small enough") are put in their own local method so the JIT can independently determine whether or not to inline using only that method (another explanation of this is here: https://source.dot.net/#System.Private.CoreLib/shared/System/Runtime/Intrinsics/Vector128.cs,11).

For example, https://source.dot.net/#System.Private.CoreLib/shared/System/Runtime/Intrinsics/Vector128.cs,312.

> Is it worth applying it to the method given the discussion around its use in tight loops

I would imagine so, the codegen is very small (5 instructions in the intrinsic case and nearly as small in the software fallback case), so the cost of the additional indirection might be greater than the cost of just running the code.

Reply:
So in this situation because both pathways are likely profitable to be inlined and the fallback is also small, just mark the whole method aggro inlining rather than separate it into a separate fallback method?

Reply:
I think it would be good to let the JIT make its own decision about the software fallback...

The code generated around the `DoubleToInt64` bits is still non-ideal (I imagine the ARM codegen is similar, but I'll need to check):
```asm
    L0000: sub rsp, 0x18
    L0004: vzeroupper
    L0007: vmovsd [rsp+0x10], xmm1
    L000d: mov rax, [rsp+0x10]
    L0012: vmovsd [rsp+0x8], xmm2
    L0018: mov rdx, [rsp+0x8]
    L001d: mov rcx, 0x8000000000000000
    L0027: and rdx, rcx
    L002a: mov rcx, 0x7fffffffffffffff
    L0034: and rax, rcx
    L0037: or rax, rdx
    L003a: mov [rsp], rax
    L003e: vmovsd xmm0, qword [rsp]
    L0043: add rsp, 0x18
    L0047: ret
```

It would also be good to try and get rid of the second constant load here, having the following would be much better:
```diff
     mov rcx, 0x8000000000000000
     and rdx, rcx
-    mov rcx, 0x7fffffffffffffff  
+    not rcx
     and rax, rcx
```

It looks like the C# compiler optimizes it to be two distinct constants in the case where `signMask` is declared a constant and the JIT does the same when it isn't.

Reply:
👍 for the separate method fallbacks. Currently trying a few things but am failing to trigger the JIT to generate the second style unfortunately

Reply:
It might be something the JIT doesn't support today, in which case logging a bug should be sufficient.